### PR TITLE
(#96) Adds Optional SonarQube Step

### PR DESCRIPTION
**Pull request by JPRuskin**
This adds build tasks for analysing and submitting results to a SonarQube instance.

It may not yet be suited to PR builds, as there are some additional settings involved, but it should suffice for scanning using the Community edition.

### CODE REVIEW

**Comment by JPRuskin** [comment]
Did you have any thoughts on this, @gep13?

**Comment by gep13** [comment]
@JPRuskin I have not had a chance to review this PR yet, sorry.

File: Chocolatey.Cake.Recipe/Content/toolsettings.cake [hunk 1]
```diff
@@ -62,6 +63,7 @@ public static class ToolSettings
         string reportUnitTool = "#tool nuget:?package=ReportUnit&version=1.2.1",
         string reSharperReportsTool = "#tool nuget:?package=ReSharperReports&version=0.2.0",
         string reSharperTools = "#tool nuget:?package=JetBrains.ReSharper.CommandLineTools&version=2017.2.0",
+        string sonarQubeTool = "#tool nuget:?package=MSBuild.SonarQube.Runner.Tool",
```
Comment (gep13) [inline Chocolatey.Cake.Recipe/Content/toolsettings.cake:66]: Does this tool work on all Operating Systems?  Or do we need to gate SonarQube execution to only run on Windows? i.e. what will happen when the tagged build runs on GitHub Actions for Mac and Linux?

**Review by gep13 — CHANGES_REQUESTED** [review]

File: Chocolatey.Cake.Recipe/Content/sonarqube.cake [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    .IsDependeeOf("Build")
+    .Does(() => RequireTool(ToolSettings.SonarQubeTool, () =>
+{
+    var SonarQubeSettings = new SonarBeginSettings
+    {
+        Key     = BuildParameters.SonarQubeId,
+        Version = BuildParameters.Version.InformationalVersion,
+        Login   = BuildParameters.SonarQubeToken
+        // Url     = "http://localhost:9000"  // Default value, works with e.g. Cloudflared tunnel
+        // Exclusions = ""
```
[the remark this answers is not in the document]
Comment (JPRuskin) [reply Chocolatey.Cake.Recipe/Content/sonarqube.cake:28]: I had left them in as examples for future potential usage, but am happy to remove them

File: Chocolatey.Cake.Recipe/Content/parameters.cake [hunk 3]
```diff
@@ -440,12 +450,28 @@ public static class BuildParameters
...
         ShouldRunReportGenerator = shouldRunReportGenerator;
         ShouldRunReportUnit = shouldRunReportUnit;
+
+        if (context.HasArgument("shouldRunSonarQube"))
+        {
+            ShouldRunSonarQube = context.Argument<bool>("shouldRunSonarQube");
+        }
+        else
+        {
+            if (BuildParameters.IsTagged && !BuildParameters.IsLocalBuild)
```
[the remark this answers is not in the document]
Comment (JPRuskin) [reply Chocolatey.Cake.Recipe/Content/parameters.cake:460]: That's correct. Our license doesn't support multiple branches per project and Paul wanted to be able to produce reports on our tagged releases. We should change this if/when we upgrade to a license that allows for, e.g., PR build scanning before merge.

[the remark this answers is not in the document]
Comment (gep13) [reply Chocolatey.Cake.Recipe/Content/parameters.cake:460]: Sounds good.

**Review by gep13 — APPROVED** [review]
LGTM!

**Pull request merged**